File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& mat) {
        vector<int> ans;
        int m = mat.size();
        int n = mat[0].size();
        int srow = 0, scol=0, erow=m-1 , ecol = n-1;
        while(srow<=erow && scol<=ecol ){
            //top
            for(int j = scol; j<=ecol;j++){
                ans.push_back(mat[srow][j]);

            }
            //right
            for(int j = srow+1; j<=erow; j++){
                ans.push_back(mat[j][ecol]);
            }
            //bottom
            for(int i = ecol-1; i>=scol;i--){
                if(srow == erow){
                    break;
                }
                ans.push_back(mat[erow][i]);
            }
            //left
            for(int i =erow-1; i>=srow+1;i-- ){
                if(scol == ecol){
                    break;
                }
                ans.push_back(mat[i][scol]);
                
            }
            srow++; erow--; scol++; ecol--;

        }
        return ans;

        
    }
};

File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        int sr = 0, er = m-1;
        int sc = 0, ec = n-1;
        ArrayList<Integer> ans = new ArrayList<>();
        while(sr<=er && sc<= ec){
            for(int j = sc; j<=ec; j++){
                ans.add(mat[sr][j]);

            }
            for(int j = sr+1; j<=er; j++){
                ans.add(mat[j][ec]);

            }
            for(int j = ec-1; j>=sc; j--){
                if(sr == er){
                    break;
                }
                ans.add(mat[er][j]);

            }
            for(int j = er-1; j>sr; j--){
                if(sc == ec){
                    break;
                }
                ans.add(mat[j][sc]);

            }
            sr++;
            er--;
            sc++;
            ec--;
        }
        return ans;
    }
}